src/gwt/panmirror/src/editor/src/behaviors/spelling/spelling-realtime.ts

```diff
@@ -196,6 +200,11 @@ class RealtimeSpellingPlugin extends Plugin<DecorationSet> {
         focusUnsubscribe();
         this.hasBeenFocused = true;
         updateSpelling(this.view);
+
+        // call a second time as no words will be cached initially, this simplifies the
+        // need for threading a callback through the entire plugin system
+        const v = this.view;
+        setTimeout(() => updateSpelling(v), 5000);
```

Comment:
Can you elaborate a bit more on why this is necessary? Is there any cleaner way we can handle sequencing here rather than a plain timeout? Or, would it be reasonable to cancel this update in some contexts?

Reply:
This is something that took me a long time to figure any decent solution to.

So basically what's happening here is there is a ProseMirror transaction that kicks off and bubbles an event through to the `spelling-realtime.ts` plugin. It's expecting a synchronous return of an array of decorations (underlines) for incorrect words. 

Since we're now async to the server we don't know the results until the RPC returns. I figured out a cute solution to this as you're typing (the `from - 1` range change so the word is already checked when you hit space/punctuation/anything after) but on the initial focus this was a bit harder. I'm totally open to any better solutions.

Early last week I had callbacks threaded from the Plugin all the way through the PanmirrorUISpelling interface. It worked but telling ProseMirror to play a transaction again caused all sorts of issues.

Reply:
A potential thought:

- `updateSpelling()` could set a flag; e.g. something like `waitingForSpellcheck_ = true`,
- The spellcheck RPC response, when received, could check that flag; if it's set, then it would send the associated "update spelling" transaction,
- `updateSpelling()` could also run the associated transaction eagerly if the local cache is up-to-date, or if explicitly requested.

I think this would require threading some more state (mainly around whether there is an update / RPC pending) but hopefully that would avoid re-wiring the existing components too much.

FWIW, my main worry is that this might in some weird contexts cause a cascade of `updateSpelling()` requests that could be noisy or do something unexpected for the user. If that's not a big concern then it might be okay leaving out of this PR?

Reply:
I haven't been able to ever trigger this more than once. I think it's pretty good about maintaining that it's ever had focus and wont trigger again later.

Threading through the callback for just this case isn't too bad. In my explanation I'm kind of conflating how I was trying to solve the "checking as you're typing" problem with this one (which doesn't have the plugin/transaction issue). I'm not too concerned with this bit of code specifically, it actually mirrors what we used to do with Typo.js pretty closely and is more conservative.